File: examples/circuit-breaker/web2/csharp/circuit-breaker-api/Controllers/CircuitBreakerController.cs
using Microsoft.AspNetCore.Mvc;

namespace circuit_breaker_api.Controllers;

[ApiController]
[Route("api")]
public class CircuitBreakerController : ControllerBase
{
    private enum BreakerState { Closed, Open, HalfOpen }

    private static BreakerState _state = BreakerState.Closed;
    private static int _failureCount = 0;
    private static DateTime _lastFailureTime = DateTime.MinValue;

    private const int Threshold = 3;
    private static readonly TimeSpan Timeout = TimeSpan.FromSeconds(10);

    [HttpGet("call")]
    public IActionResult CallService()
    {
        var now = DateTime.UtcNow;

        if (_state == BreakerState.Open && (now - _lastFailureTime > Timeout))
        {
            _state = BreakerState.HalfOpen;
        }

        if (_state == BreakerState.Open)
        {
            return StatusCode(503, new { error = "Circuit is OPEN" });
        }

        // Simulate random failure
        if (Random.Shared.NextDouble() > 0.5) // Success
        {
            _state = BreakerState.Closed;
            _failureCount = 0;
            return Ok(new { status = "success" });
        }
        else // Failure
        {
            _failureCount++;
            _lastFailureTime = now;

            if (_state == BreakerState.HalfOpen || _failureCount >= Threshold)
            {
                _state = BreakerState.Open;
            }

            return StatusCode(500, new { status = "fail", state = _state.ToString() });
        }
    }
}

File: examples/circuit-breaker/web2/springboot/src/main/java/com/example/demo/CircuitBreakerController.java
package com.example.demo;

import org.springframework.web.bind.annotation.*;
import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;
import java.util.concurrent.atomic.AtomicLong;

@RestController
@RequestMapping("/api")
public class CircuitBreakerController {

    private enum State { CLOSED, OPEN, HALF_OPEN }
    
    private State state = State.CLOSED;
    private final AtomicInteger failureCount = new AtomicInteger(0);
    private final AtomicLong lastFailureTime = new AtomicLong(0);
    
    private static final int THRESHOLD = 3;
    private static final long TIMEOUT = 10000; // 10s

    @GetMapping("/call")
    public synchronized ResponseEntity<?> callService() {
        long now = System.currentTimeMillis();

        if (state == State.OPEN && (now - lastFailureTime.get() > TIMEOUT)) {
            state = State.HALF_OPEN;
        }

        if (state == State.OPEN) {
            return ResponseEntity.status(HttpStatus.SERVICE_UNAVAILABLE)
                    .body(Map.of("error", "Circuit is OPEN"));
        }

        if (Math.random() > 0.5) { // Success
            state = State.CLOSED;
            failureCount.set(0);
            return ResponseEntity.ok(Map.of("status", "success"));
        } else { // Failure
            int currentFailures = failureCount.incrementAndGet();
            lastFailureTime.set(now);
            
            if (state == State.HALF_OPEN || currentFailures >= THRESHOLD) {
                state = State.OPEN;
            }
            return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR)
                    .body(Map.of("status", "fail", "state", state));
        }
    }
}
